core/src/rpc_subscriptions.rs

```diff
@@ -60,44 +61,49 @@ impl std::fmt::Debug for NotificationEntry {
     }
 }
 
+struct SubscriptionData<S> {
+    sink: Sink<S>,
+    commitment: CommitmentConfig,
+    last_notified_slot: RwLock<Slot>,
+}
 type RpcAccountSubscriptions =
-    RwLock<HashMap<Pubkey, HashMap<SubscriptionId, (Sink<Response<RpcAccount>>, Confirmations)>>>;
-type RpcProgramSubscriptions = RwLock<
-    HashMap<Pubkey, HashMap<SubscriptionId, (Sink<Response<RpcKeyedAccount>>, Confirmations)>>,
->;
+    RwLock<HashMap<Pubkey, HashMap<SubscriptionId, SubscriptionData<Response<RpcAccount>>>>>;
+type RpcProgramSubscriptions =
+    RwLock<HashMap<Pubkey, HashMap<SubscriptionId, SubscriptionData<Response<RpcKeyedAccount>>>>>;
 type RpcSignatureSubscriptions = RwLock<
-    HashMap<
-        Signature,
-        HashMap<SubscriptionId, (Sink<Response<RpcSignatureResult>>, Confirmations)>,
-    >,
+    HashMap<Signature, HashMap<SubscriptionId, SubscriptionData<Response<RpcSignatureResult>>>>,
 >;
 type RpcSlotSubscriptions = RwLock<HashMap<SubscriptionId, Sink<SlotInfo>>>;
 type RpcRootSubscriptions = RwLock<HashMap<SubscriptionId, Sink<Slot>>>;
 
 fn add_subscription<K, S>(
-    subscriptions: &mut HashMap<K, HashMap<SubscriptionId, (Sink<S>, Confirmations)>>,
+    subscriptions: &mut HashMap<K, HashMap<SubscriptionId, SubscriptionData<S>>>,
     hashmap_key: K,
-    confirmations: Option<Confirmations>,
+    commitment: Option<CommitmentConfig>,
     sub_id: SubscriptionId,
     subscriber: Subscriber<S>,
 ) where
     K: Eq + Hash,
     S: Clone,
 {
     let sink = subscriber.assign_id(sub_id.clone()).unwrap();
-    let confirmations = confirmations.unwrap_or(0);
-    let confirmations = min(confirmations, MAX_LOCKOUT_HISTORY + 1);
+    let commitment = commitment.unwrap_or_else(CommitmentConfig::recent);
+    let subscription_data = SubscriptionData {
+        sink,
+        commitment,
+        last_notified_slot: RwLock::new(0),
```

Comment:
Because `last_notified_slot` is initialized as 0, if a subscription is made for an account that already has some changes in the AccountsDB, it will receive a notification right away, marked with the last change slot.
This seems okay to me. The alternative involves making a Bank/AccountsDB available to `add_account_subscription`; probably a bunch of churn, but doable.

Reply:
I think that behaviour would be unexpected. I think clients would expect to only receive updates that take place after their subscription. It would be weird to immediately get a notification for data that changed a month ago.

Reply:
We could have logic to update `last_notified_slot` but not notify if it is 0 🤔

Also, should it be `last_checked_slot`?

Reply:
I don't think that will work for an account that is empty when you subscribe to it, but subsequently receives an update; it would not notify. I'll think on this.

Reply:
Okay, updated to properly initialize `last_notified_slot`

Regarding the name of that variable, I agree that `last_notified` isn't perfect, but I think it's closer than `last_checked`. It's not necessarily the last slot that was checked for changes, but the last slot that had changes that triggered a notification. I'll try to think of something better.

Reply:
Yeah, `last_checked` isn't right at all. `last_notified` seems fine 😄